### PR TITLE
Rework yarpcerrors API

### PR DESCRIPTION
**Pull request by peter-edge**
@breerly per our conversation last night

This is what I'm proposing. This effectively takes the API down to four functions:

1. `Newf`: This is how you create errors. It returns a `*Status`.
2. `FromError`: This is how you convert generic errors to `*Status`es. If the given error is a `*Status`, it returns it, otherwise it returns a `*Status` with `CodeUnknown` and message `err.Error()`.
3. `IsStatus`: This checks if an error is a `*Status`. This is and `FromError` replace grpc-go's `FromError` https://godoc.org/google.golang.org/grpc/status#FromError, I think the general case here is better instead of making everyone do their own conversion. If you really care if the error is not a `*Status`, you can check that (we need `IsStatus` for metrics for example).
4. `Status#WithName`: This adds a user-defined name to the `*Status`. This takes away the restriction of only having `CodeUnknown` associated with a named error. We could even think of this as the simplest version of `WithDetails`.

`*Status` has the functions `Code, Name, Message` and also `Error` to implement `error`.

Most users should only have to do two things:

1. Call `yarpcerrors.Newf` to create a YARPC error.
2. Call `yarpcerrors.FromError` to handle a YARPC error.

This only ends up deprecating three functions:

1. `IsYARPCError`: This is just effectively renamed to `IsStatus`.
2. `NamedErrorf`: This is replaced by using `Newf` and `WithName`.
3. `FromHeaders`: This is replaced by using `Newf` and `WithName`.

We don't even need to deprecate all the helpers, although we could mark them as deprecated just to encourage using only the above functions.

This is completely backwards-compatible and all tests pass, except linting for now - linting fails because I still use the deprecated functions, I didn't want to go through and change them until we agree on the final API.

### CODE REVIEW

**Comment by breerly** [comment]
I'm digging how this feels from the end-user's POV - thanks for jamming yesterday as I tried to document this stuff.

Can you briefly touch on:

A. How this API could lead to nicely annotating Thrift exceptions with their correct code.
B. How middleware could determine which code was thrown, e.g. signal SRE/circuit breaking that the service is falling over.
C. How to know if a code is retryable.
D. How this leads to being able to support something like gRPC's error detail approach.

**Comment by peter-edge** [comment]
```
A. How this API could lead to nicely annotating Thrift exceptions with their correct code.
B. How middleware could determine which code was thrown, e.g. signal SRE/circuit breaking that the service is falling over.
C. How to know if a code is retryable.
```

This PR doesn't touch on these subjects, they are all orthogonal to what is going on here. (A) will be done in another PR, (B) and (C) are unchanged.

```
D. How this leads to being able to support something like gRPC's error detail approach.
```

A function will be added later to `*Status`:

```go
func (s *Status) WithDetails(details interface{}) *Status { ... }
```

File: yarpcerrors/errors.go [hunk 1]
```diff
@@ -25,24 +25,112 @@ import (
...
+		message: sprintf(format, args...),
+	}
+}
+
+// FromError returns the Status for the error.
+//
+// If the error is a Status, it returns err.
+// If the error is not a Status, it returns a new error with the Code
+// CodeUnknown and the message err.Error()
+func FromError(err error) *Status {
```
Comment (willhug) [inline yarpcerrors/errors.go]: add tests plz

File: yarpcerrors/errors.go [hunk 2]
```diff
@@ -25,24 +25,112 @@ import (
...
+// WithName returns a new Status with the given name.
+//
+// If s is nil, this will still return nil.
+//
+// This should be used for user-defined errors.
+//
+// The name must only contain lowercase letters from a-z and dashes (-), and
+// cannot start or end in a dash. If the name is something else, an error with
+// code CodeInternal will be returned.
+func (s *Status) WithName(name string) *Status {
```
Comment (willhug) [inline yarpcerrors/errors.go]: do we want an equivalent "WithCode"

My thought is something like:

`yarpcerrors.FromError(err).WithCode(yarpcerrors.CodeInvalidRequest)`

**Comment by willhug** [comment]
also the test plz

**Comment by breerly** [comment]
This proposal looks great for me - @peter-edge lets add tests and get this out of WIP.

I'm going to update our end-user guide to be based off this 👍

File: yarpcerrors/errors.go (hunk 2 above)
Comment (peter-edge) [reply yarpcerrors/errors.go]: I think we can do that after, let's just have the basics now

File: yarpcerrors/errors.go (hunk 1 above)
Comment (peter-edge) [reply yarpcerrors/errors.go]: Yea, going go through through everything and update it

**Comment by peter-edge** [comment]
I've finished the work on this, this is ready to go. I converted all of yarpc-go to use the new API as well, which was mostly just substitute/replace. I did this by marking all existing functions as deprecated, and then having `staticcheck` tell me where I was calling deprecated functions.

The question I have is: do we want to leave all the helpers as deprecated? I want our internal code not to use them, but so we want to leave the following functions for users, or get users into the habit of using the smaller API in case we want to delete them in 2.0?

- `ErrorCode`
- `ErrorName`
- `ErrorMessage`
- `IsInvalidArgument, IsNotFound, et`
- `InvalidArgumentErrorf, NotFoundErrorf, etc`

I think they are actually easier to use in some cases, but I don't know, they could just confuse users, and I wish the `*Errorf` functions returned `*Status`. I could go either way. As of now they are all marked as deprecated.

**Comment by peter-edge** [comment]
Also note the `yarpcerrors` package now has 100% code coverage by itself (ie without cross-package coverage testing) @willhug 

```
$ go test -cover ./yarpcerrors
ok  	go.uber.org/yarpc/yarpcerrors	0.081s	coverage: 100.0% of statements
```

**Comment by breerly** [comment]
@peter-edge and yeah, I think leaving the helpers as deprecated makes sense: less is more, and we can't delete them just yet either.

**Comment by peter-edge** [comment]
Hmm, my personal opinion would be to land it now and just work to get the guide up today - I want to do a release for #1295 and I don't want to have to do a second release for this, and right now there's really only one customer using this so I think it's harmless. Is that OK?

**Comment by peter-edge** [comment]
Also I wanted to merge it so I don't have to chain some work on the thrift exception code mapping thing.

**Comment by breerly** [comment]
:thumbs:

**Pull request merged**